### PR TITLE
Mypy on parsec.api

### PR DESCRIPTION
**Pull request by AntoineDupre**
Add mypy check

### CODE REVIEW

File: mypy.ini [hunk 1]
```diff
@@ -13,11 +13,45 @@ plugins = trio_typing.plugin
...
 disallow_any_decorated=True
 disallow_any_explicit=True
 disallow_any_generics=True
 disallow_subclassing_any=True
 
 [mypy-parsec.core.fs.workspacefs.versioning_helpers]
 disallow_untyped_defs=False
+
+
+[mypy-parsec.api.protocol.*]
```
Comment (vxgmichel) [inline mypy.ini:25]: No common configuration for `parsec.api`?

File: parsec/api/data/certif.py [hunk 2]
```diff
@@ -145,20 +150,23 @@ class SCHEMA_CLS(BaseSignedDataSchema):
...
     device_label: Optional[str]
     verify_key: VerifyKey
 
     @classmethod
     def verify_and_load(
-        cls, *args, expected_device: Optional[DeviceID] = None, **kwargs
+        cls: Type[DeviceCertificateContentTypeVar],
+        *args: Any,
+        expected_device: Optional[DeviceID] = None,
+        **kwargs,
```
Comment (vxgmichel) [inline parsec/api/data/certif.py:169]: `kwargs` not typed here?

File: parsec/api/data/invite.py [hunk 3]
```diff
@@ -13,21 +13,24 @@
...
         if not isinstance(raw, str) or not self.regex.match(raw):
             raise ValueError("Invalid SAS code")
 
-    def __repr__(self):
+    def __repr__(self) -> str:
         return f"<SASCode {super().__repr__()}>"
 
     @classmethod
-    def from_int(cls, num):
+    def from_int(cls: Type[SASCodeTypeVar], num: int) -> SASCodeTypeVar:
```
Comment (vxgmichel) [inline parsec/api/data/invite.py:33]: Why not simply `def from_int(cls, num: int) -> "SASCode":`?

File: parsec/api/data/manifest.py [hunk 4]
```diff
@@ -200,17 +214,32 @@ class SCHEMA_CLS(BaseSignedDataSchema):
...
 
+    @classmethod
+    def verify_and_load(  # type: ignore[override]
+        cls: Type["FileManifest"],
+        *args: object,
+        expected_parent: Optional[EntryID] = None,
+        **kwargs: object,
+    ) -> "FileManifest":
+        data = super().verify_and_load(*args, **kwargs)  # type: ignore[arg-type]
+        print(data)
```
Comment (vxgmichel) [inline parsec/api/data/manifest.py:236]: Maybe run a `git grep print`?

Comment (vxgmichel) [inline parsec/api/data/manifest.py:236]: Ooopsie

File: parsec/api/protocol/types.py [hunk 5]
```diff
@@ -17,58 +22,58 @@ class OrganizationID(str):
...
+    def __repr__(self) -> str:
         return f"<UserID {super().__repr__()}>"
 
     @classmethod
-    def new(cls):
+    def new(cls: Type[UserIDTypeVar]) -> UserIDTypeVar:
         return cls(uuid4().hex)
 
-    def to_device_id(self, device_name: Union[str, "DeviceName"]) -> "DeviceID":
+    def to_device_id(self, device_name: Union[str, DeviceNameTypeVar]) -> "DeviceID":
```
Comment (vxgmichel) [inline parsec/api/protocol/types.py:48]: Again, was a TypeVar really necessary here?

**Review by vxgmichel — APPROVED** [review]
Great job, that must have been tedious ! :+1:

File: mypy.ini (hunk 1 above)
Comment (AntoineDupre) [reply mypy.ini]: fixed

File: parsec/api/data/certif.py (hunk 2 above)
Comment (AntoineDupre) [reply parsec/api/data/certif.py]: fixed

File: parsec/api/data/invite.py (hunk 3 above)
Comment (AntoineDupre) [reply parsec/api/data/invite.py]: fixed

File: parsec/api/protocol/types.py (hunk 5 above)
Comment (AntoineDupre) [reply parsec/api/protocol/types.py]: fixed

**Pull request merged**